cognite/client/_api/data_modeling/instances.py

```diff
@@ -394,6 +414,80 @@ def delete(
+
+        We do not support chaining result sets when subscribing to a query.
+
+        Args:
+            query (Query): The query to subscribe to.
+            callback (Callable[[QueryResult], None]): The callback function to call when the result set changes.
+            max_poll_backoff_seconds (float): The maximum time to wait between polls. Defaults to 30 seconds.
+        Returns:
+            SubscriptionContext: An object that can be used to cancel the subscription.
+
+        Examples:
+
+            Delete instances by id:
+
+                >>> from cognite.client import CogniteClient
+                >>> from cognite.client.data_classes.data_modeling.query import Query, QueryResult, NodeResultSetExpression, Select, SourceSelector
+                >>> from cognite.client.data_classes.data_modeling import ViewId
+                >>> from cognite.client.data_classes.filters import Range
+                >>>
+                >>> c = CogniteClient()
+                >>> def just_print_the_result(result: QueryResult) -> None:
+                ...     print(result)
+                ...
+                >>> view_id = ViewId("someSpace", "someView", "v1")
+                >>> filter = Range(view_id.as_property_ref("releaseYear"), lt=2000)
+                >>> query = Query(
+                ...     with_={"movies": NodeResultSetExpression(filter=filter)},
+                ...     select={"movies": Select([SourceSelector(view_id, ["releaseYear"])])}
+                ... )
+                >>> subscription_context = c.data_modeling.instances.subscribe(query, just_print_the_result)
+                >>> subscription_context.cancel()
+        """
+        for result_set_expression in query.with_.values():
+            if result_set_expression.from_ is not None:
+                raise ValueError("Cannot chain result sets when subscribing to a query")
+
+        subscription_context = SubscriptionContext()
+
+        def _do_subscribe() -> None:
+            cursors = query.cursors
+            backoff = Backoff(max_wait=max_poll_backoff_seconds)
+            while not subscription_context._canceled:
+                try:
+                    query.cursors = cursors
+                    result = self.sync(query)
+                    subscription_context.last_successful_sync = datetime.now(tz=timezone.utc)
+                    if any(len(instances) > 0 for instances in result.data.values()):
+                        callback(result)
+                        subscription_context.last_successful_callback = datetime.now(tz=timezone.utc)
+                        # only progress the cursor if the callback executed successfully
+                        cursors = result.cursors
+                        backoff.reset()
+                        continue
+                    else:
+                        # No data was returned, so we back off before we try again.
+                        time.sleep(next(backoff))
```

Comment:
I think polling delay/interval and error backoff should be separate concepts.

If there's a poll interval, I'd delay the first one with `random.uniform(0, poll_interval)` to smear out a bit, but let the subsequent ones go with a consistent delay.

If `self.sync` throws, we shouldn't need to re-query before the next attempt.

Reply:
Sorry, I meant if `callback` throws.